dataikuapi/dss/ml.py

```diff
@@ -315,43 +313,979 @@ def save(self):
+        """
+        :param values: the explicit list of numerical values considered for this hyperparameter in the search
+        :type values: list of float | int
+        """
+        error_message = "Invalid values input type for hyperparameter " \
+                        "\"{}\": ".format(self.name) + \
+                        " expecting a non-empty list of numbers"
+        assert values is not None and isinstance(values, list) and len(values) > 0, error_message
+        for val in values:
+            assert isinstance(val, int) or isinstance(val, float), error_message
+        limit_min = self._algo_settings[self.name]["limit"].get("min")
+        if limit_min is not None:
+            assert all(limit_min <= val for val in values), "Value(s) below hyperparameter \"{}\" limit {}".format(self.name, limit_min)
+        limit_max = self._algo_settings[self.name]["limit"].get("max")
+        if limit_max is not None:
+            assert all(val <= limit_max for val in values), "Value(s) above hyperparameter \"{}\" limit {}".format(self.name, limit_max)
+        if len(set(values)) < len(values):
+            warnings.warn("Detected duplicates in provided values: " + str(sorted(values)))
+        self._algo_settings[self.name]["values"] = values
+
+    def _check_number_input(self, input):
+        assert isinstance(input, int) or isinstance(input, float), \
+            "Invalid input type for hyperparameter \"{}\": ".format(self.name) + \
+            "range bounds must be numbers"
+
+    def _set_range(self, min=None, max=None, nb_values=None):
+        if min is None and max is None and nb_values is None:
+            warnings.warn("Numerical range for hyperparameter \"{}\" not modified".format(self.name))
+        else:
+            if min is not None:
+                self._check_number_input(min)
+                limit_min = self._algo_settings[self.name]["limit"].get("min")
+                if limit_min is not None:
+                    assert limit_min <= min, "Range min {} is below hyperparameter \"{}\" limit {}".format(min, self.name, limit_min)
+            if max is not None:
+                self._check_number_input(max)
+                limit_max = self._algo_settings[self.name]["limit"].get("max")
+                if limit_max is not None:
+                    assert max <= limit_max, "Range max {} is above hyperparameter \"{}\" limit {}".format(max, self.name, limit_max)
+            if nb_values is not None:
+                assert isinstance(nb_values, int) and nb_values >= 2, "Range number of values for hyperparameter \"{}\" must be an integer and >= 2".format(self.name)
+
+            if min is not None:
+                self._algo_settings[self.name]["range"]["min"] = min
+            if max is not None:
+                self._algo_settings[self.name]["range"]["max"] = max
+            if nb_values is not None:
+                self._algo_settings[self.name]["range"]["nbValues"] = nb_values
```

Comment:
We do use variables with names `min` & `max`, that are overriding builtin functions. This does not break the code but I think it is not a good practice.

Besides, we do a lot of checks but we don't check that `min <= max`, if they are both defined

Reply:
Using `min` and `max` naming offers the benefit of transparency (and arguably simplicity) to the user with the downside of a not-so-great implementation. The only real risk I can see with this is encouraging the user to use variables `min` and `max` themselves and then override the builtin functions, which could cause trouble down the line.
Using something like `maximum` and `minimum` as kwargs would not prevent this though... and would add a bit of complexity by introducing a mismatch between the json and the python object, so it's not clear to me which one is best for the user. Having the internal implementation not 100% clean (overriding `min` and `max` builtin functions in a scope where where is no chance they are being uses) is not a big deal IMO if it is better for the user, but the question is: is it ?

Regardless, we should probably check that `min <= max` indeed !

Reply:
Not a big deal indeed, I'd favor direct mapping to the JSON for the user's sake.

Reply:
let's keep `min` & `max` then